Feature/Topology discovery - trigger topology query message

pavellevitsky opened the pull request:
- topology discovery class declaration (yaml files will create autogenerated header and cpp files)
- add topology query to beerocks cli bml commands map
- declare and implement internal topology discovery function

#120

Commit:
common: extend yaml files with Topology Query class

In yaml files new topology discovery definitions added
In control header files beerocks cli bml commands map
is extended by topology discovery.

Signed-off-by: pavel levitsky <pavellev@inter.net.il>

tomereli reviewed (changes requested):
1. Please run clang-format
2. Please rebase
3. Please update the commit message headline. See the CONTRIBUTING.md for instructions on how to write a commit message.
4. This looks like a part of feature - why not create a PR with multiple commits once the feature is actually ready?

tomereli commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -512,6 +512,13 @@ int bml_wfca_controller(BML_CTX ctx, const char *cmd, char *ret_buf, int ret_buf
+ *
+ * @return BML_RET_OK on success.
+ */
+int bml_topology_discovery(void);
```
please change to `bml_trigger_topology_discovery(void)`

tomereli reviewed (changes requested):
I am still missing the actual implementation - where is the controller changes? where is the creation and send of the CMDU? These should all be in this PR.

Commit:
controller: build and send CMDU Topology Query message

BML: new function to call new bml_internal function that sends topology query request to controller.
CLI: new beerocks_cli command to call BML API (above), its argument(s) validator and caller.
Master: implement message handler for topology query request sent from BML.

Signed-off-by: pavel levitsky <pavellev@inter.net.il>

tomereli commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -512,6 +512,16 @@ int bml_wfca_controller(BML_CTX ctx, const char *cmd, char *ret_buf, int ret_buf
+ *
+ * @return BML_RET_OK on success.
+ */
+int bml_trigger_topology_discovery(BML_CTX ctx, const char *al_mac);
```
input parameter as char* without specifying the length and doing sanity that it ends with '\0' is risky and prone for errors.
Since the user enters the MAC address as string in the CLI, please convet it to `sMacAddr` there and change all of the APIs to accept `sMacAddr`

Galvien replied:
keeping as is since the conversion to sMacAddr happens in `bml_internal` and the actual char * is received from std::string which is safe since it adds the `\0`.

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1287,6 +1291,14 @@ int cli_bml::bml_wfca_agent_caller(int numOfArgs)
+int cli_bml::bml_trigger_topology_discovery_caller(int numOfArgs)
+{
+    if (numOfArgs == 1) {
+        return topology_discovery(args.stringArgs[0]);
```
no verification for user argument, potential security breach. Need to convert to `sMacAddr` and if fails return error. Then pass the struct instead of const char *.

Galvien replied:
see previous comment

tomereli reviewed (changes requested):
Other than the issue with passing char * without length to a C function - looks good.
My suggestion is to use `mac_from_string()` on the user input (make sure it does verify it) and pass the `sMacAddr` structure from here on.

In addition, please add minimal implementation at the agent to only print that it received the topology query - I guess it should be in the main agent log.
Then we can merge this and add another task to add this to CI.

arnout commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1720,6 +1720,27 @@ void son_management::handle_bml_message(
+
+        auto bml_request = cmdu_rx.addClass<beerocks_message::cACTION_BML_TRIGGER_TOPOLOGY_QUERY>();
+
+        std::string al_mac = network_utils::mac_to_string(bml_request->al_mac());
```
In general, we prefer to use `auto`.

arnout commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -1720,6 +1720,27 @@ void son_management::handle_bml_message(
+            return;
+        }
+
+        Socket *agent_sd = database.get_node_socket(al_mac);
```
You need to check the return value.

@tomereli BTW it is not clear to me if we should use `!agent_sd` (using implicit conversion to bool) or `== std::nullptr` (using explicit operator ==).

Another comment (not for Pavel): I think the database should be converted to use sMacAddr instead of std::string for the MAC address lookups. Should I make a refactoring issue for that?

tomereli replied:
`!agent_sd` should be used as far as I remember to avoid KW issues.
Regarding refactoring - by all means, go ahead:)

arnout reviewed (commented):
(nitpick) comments only, the important thing to change is in Tomer's review.

Also a few suggestions for the future:

* I prefer that adding a new TLV struct to YAML is really a separate commit, without anything else (now you have changes in `bml.h` etc. in the same file).
* Please change your git config to correctly capitalize your name.
* Use empty lines to separate the paragraphs in your commit message.

Commit:
agent: add topology query handler stub

Add topology discovery handler which simply prints that a topology
query message was received. This will be used by a test script
which will be added in a followup commit that will trigger topology query
message from CLI via BML in the controller, then connect to the
repeater container and grep the log.
Finally it will allow testing that topology query test can pass in controller
certification.

Signed-off-by: Gal Wiener <gal.wiener@intel.com>

tomereli commented:
LGTM
